### ISSUE TITLE
tolist called on torch scalar does not return a list

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#52262]
I think it's better if it returned a single-element list instead. This matches more `tolist()` method name and makes return type more reliable.

```python
import torch
torch.tensor(1).tolist()
# 1
```

**Comment by rgommers** [comment]
I agree that a length-1 list would be better. It was on purpose though, the docstring:
```
Returns the tensor as a (nested) list. For scalars, a standard
Python number is returned, just like with :meth:`~Tensor.item`.
```
So I'm not sure we should be making a backwards-incompatible change here, may not be worth it.

**Comment by vadimkantorov** [comment]
Variable return type is quite inconsistent :( Maybe even clients that relied on it had to hack around this.

**Comment by gchanan** [comment]
This is the same behavior as numpy.  The description isn't great; the NumPy one is better:

> If a.ndim is 0, then since the depth of the nested list is 0, it will not be a list at all, but a simple Python scalar.

another way of thinking about this is you need to do this if you want the operation to be invertible (i.e. to pass it back via `torch.tensor` and get the same thing).

**Comment by vadimkantorov** [comment]
Maybe worth adding an argument that would guarantee returning a list... I had this when DataParallel automatically converted concatenated scalars to a tensor. And then logging them has to check whether it's a scalar or a larger tensor.... Basically no easy way to iterate the result of such DP concatenating for logging

**Comment by mruberry** [comment]
I appreciate your frustration, @vadimkantorov, a function named tolist() doesn't always return a list! As a general philosophy I also dislike when functions don't consistently return the same type.

However, because this is consistent with NumPy (as @gchanan notes) and the change would be BC-breaking (as @rgommers) notes, plus the workaround is somewhat straightforward (although it's also frustrating that it's not just `list(t.tolist())`) it's unlikely we'll pursue this or accept a PR for it.

Looking at some usage examples it seems people are mostly using tolist() in unambiguous circumstances where they know they're not dealing with a scalar tensor.

**Comment by ZhiyuanChen** [comment]
@mruberry 
> Looking at some usage examples it seems people are mostly using tolist() in unambiguous circumstances where they know they're not dealing with a scalar tensor.

NO. I use `tolist()` to aggregate all outputs of model by batch to a list to compute scores such as AUC. This will leads to an exception when the final batch has only one element. `drop_last` could resolve this issue to some extent, however, this damage the confidence of the result. 

Normally it is as easy as `inferences.extend(outputs.tolist())`, however, I will have to write many additional code to determine the type of `tolist()` as it would raise a `TypeError` if the result is a `float`.
